### ISSUE TITLE
TypeError when re-saving EDAX files to .hspy format

### ISSUE THREAD

**Issue by jat255** [hyperspy/hyperspy#1916]
I came across the following error when working with a lazy EDS dataset. The data was loaded from an `.hspy` file lazily, then cropped a bit, computed into memory, and then saved:

```python
>>> s = hs.load('pyrite.hspy')
>>> crop = s.inav[:,:,::5].isig[:2.6]
>>> crop.compute()
>>> crop.save('pyrite_50frames_2_6keV.hspy')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in overwrite_dataset(group, data, key, signal_axes, **kwds)
    548 
--> 549             dset = group.require_dataset(key, **these_kwds)
    550             got_data = True

C:\anaconda3\envs\hyperspy\lib\site-packages\h5py\_hl\group.py in require_dataset(self, name, shape, dtype, exact, **kwds)
    126             if not name in self:
--> 127                 return self.create_dataset(name, *(shape, dtype), **kwds)
    128 

C:\anaconda3\envs\hyperspy\lib\site-packages\h5py\_hl\group.py in create_dataset(self, name, shape, dtype, data, **kwds)
    105         with phil:
--> 106             dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
    107             dset = dataset.Dataset(dsid)

C:\anaconda3\envs\hyperspy\lib\site-packages\h5py\_hl\dataset.py in make_new_dset(parent, shape, dtype, data, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times)
     99             dtype = numpy.dtype(dtype)
--> 100         tid = h5t.py_create(dtype, logical=1)
    101 

h5py\h5t.pyx in h5py.h5t.py_create()

h5py\h5t.pyx in h5py.h5t.py_create()

h5py\h5t.pyx in h5py.h5t.py_create()

TypeError: No conversion path for dtype: dtype('<U32')

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-12-c5592dab8e1f> in <module>()
----> 1 small_crop.save('pyrite_50frames_2_6keV.hspy')

c:\users\jat\git_repos\hyperspy\hyperspy\signal.py in save(self, filename, overwrite, extension, **kwds)
   2133             basename, ext = os.path.splitext(filename)
   2134             filename = basename + '.' + extension
-> 2135         io.save(filename, self, overwrite=overwrite, **kwds)
   2136 
   2137     def _replot(self):

c:\users\jat\git_repos\hyperspy\hyperspy\io.py in save(filename, signal, overwrite, **kwds)
    480             overwrite = overwrite_method(filename)
    481         if overwrite is True:
--> 482             writer.file_writer(filename, signal, **kwds)
    483             _logger.info('The %s file was created' % filename)
    484             folder, filename = os.path.split(os.path.abspath(filename))

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in file_writer(filename, signal, *args, **kwds)
    729             smd.record_by = ""
    730         try:
--> 731             write_signal(signal, expg, **kwds)
    732         except BaseException:
    733             raise

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in write_signal(signal, group, **kwds)
    687     original_par = group.create_group(original_metadata)
    688     dict2hdfgroup(signal.original_metadata.as_dictionary(), original_par,
--> 689                   **kwds)
    690     learning_results = group.create_group('learning_results')
    691     dict2hdfgroup(signal.learning_results.__dict__,

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in dict2hdfgroup(dictionary, group, **kwds)
    434         if isinstance(value, dict):
    435             dict2hdfgroup(value, group.create_group(key),
--> 436                           **kwds)
    437         elif isinstance(value, DictionaryTreeBrowser):
    438             dict2hdfgroup(value.as_dictionary(),

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in dict2hdfgroup(dictionary, group, **kwds)
    434         if isinstance(value, dict):
    435             dict2hdfgroup(value, group.create_group(key),
--> 436                           **kwds)
    437         elif isinstance(value, DictionaryTreeBrowser):
    438             dict2hdfgroup(value.as_dictionary(),

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in dict2hdfgroup(dictionary, group, **kwds)
    434         if isinstance(value, dict):
    435             dict2hdfgroup(value, group.create_group(key),
--> 436                           **kwds)
    437         elif isinstance(value, DictionaryTreeBrowser):
    438             dict2hdfgroup(value.as_dictionary(),

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in dict2hdfgroup(dictionary, group, **kwds)
    434         if isinstance(value, dict):
    435             dict2hdfgroup(value, group.create_group(key),
--> 436                           **kwds)
    437         elif isinstance(value, DictionaryTreeBrowser):
    438             dict2hdfgroup(value.as_dictionary(),

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in dict2hdfgroup(dictionary, group, **kwds)
    443             write_signal(value, group.require_group(kn))
    444         elif isinstance(value, (np.ndarray, h5py.Dataset, da.Array)):
--> 445             overwrite_dataset(group, value, key, **kwds)
    446         elif value is None:
    447             group.attrs[key] = '_None_'

c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in overwrite_dataset(group, data, key, signal_axes, **kwds)
    552             # if the shape or dtype/etc do not match,
    553             # we delete the old one and create new in the next loop run
--> 554             del group[key]
    555     if dset == data:
    556         # just a reference to already created thing

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

C:\anaconda3\envs\hyperspy\lib\site-packages\h5py\_hl\group.py in __delitem__(self, name)
    301     def __delitem__(self, name):
    302         """ Delete (unlink) an item from this group. """
--> 303         self.id.unlink(self._e(name))
    304 
    305     @with_phil

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\h5g.pyx in h5py.h5g.GroupID.unlink()

KeyError: "Couldn't delete link (callback link pointer is NULL (specified link may be '.' or not exist))"
```

It doesn't appear to break saving however, because I can load the resulting file and the data is all the same as the original. Loading and then saving the resulting file works fine, so this isn't a showstopper, but it does present a pretty scary-looking warning that your data might not have been saved. 

I'm not really familiar with the internals of `h5py`, but I can look into it a bit more, if no one else know what's going on off-hand.

**Comment by jat255** [comment]
Actually, this also just happened with a file that was not loaded lazily, but had the same processing done otherwise. Hmm...

**Comment by ericpre** [comment]
Could it be possible to make a minimal example reproducing the issue?

**Comment by jat255** [comment]
I just ran into this again and it looks like it might be a symptom of [this](https://github.com/h5py/h5py/issues/289) issue. My guess is that it might be because of the unsupported unicode type that is present in the original metadata (as read from the EDAX format), which would explain why it only shows sometimes (and only for me, since I don't think too many people are using the EDAX reader).

**Comment by jat255** [comment]
So it looks like upon saving an EDAX .spd file to HDF5, the `charText` attribute of the `ipr_header` metadata is converted from `S32` dtype to `U32`. When this file is then loaded, all is well, but if it is re-saved to `.hspy` format, the `h5py` library doesn't know what to do with the unicode dtype.

See the following example:
```python
>>> test = hs.load('../edax_data/srf2/map20180410172803301_0_frame001.spd', 
               spc_fname='../edax_data/srf2/map20180410172803301_0.spc', 
               ipr_fname='../edax_data/srf2/map20180410172803301_0_Img.ipr')
>>> print(test.original_metadata.ipr_header.charText, test.original_metadata.ipr_header.charText.dtype)
>>> test.save('test.hspy', overwrite=True)
>>> test_load = hs.load('test.hspy')
>>> print(test_load.original_metadata.ipr_header.charText, test_load.original_metadata.ipr_header.charText.dtype)
>>> test_load.save('test_save.hspy', overwrite=True)
```
```
[b'' b'' b'' b''] |S32
['' '' '' ''] <U32
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
c:\users\jat\git_repos\hyperspy\hyperspy\io_plugins\hspy.py in overwrite_dataset(group, data, key, signal_axes, **kwds)
    548 
--> 549             dset = group.require_dataset(key, **these_kwds)
    550             got_data = True

... etc ...
```

I see two options:
1) Explicitly encode the `charText` attribute to some encoding (like UTF-8)
2) Just blank out `charText` upon loading, so we don't have to deal with it

Although I am hesitant to "remove functionality", I have not yet come across an .ipr file thus far that has any data in the `charText` tag, so I don't think EDAX is using it. I would have no problem going with option 2 to just avoid future incarnations of this problem.

**Comment by ericpre** [comment]
Fixed in #2007.
